### PR TITLE
Nano: Add CUDA unpatch

### PR DESCRIPTION
**Pull request by MeouSker77**
## Description

- Add `unpatch_cuda()`, which will be called automatically in `unpatch_torch()`.
- Protect the patch and unpatch from being called more than once.
- Eliminate the usage of `BIGDL_NANO_PATCH_TORCH` and `BIGDL_NANO_PATCH_CUDA` environment variable.

### 1. Why the change?

The `unpatch_torch()` function will not unpatch CUDA's patch for now, so we fix it in this PR.

### 2. User API changes

N/A

### 3. Summary of the change 

- Replace `setattr` with `replace_attr` helper function, which will not only set new attribute, but also store the original attributes, then use them to do unpatch.
- Use global variables instead of environment variables to indicate the status of patch

### 4. How to test?
- [ ] N/A
- [x] Unit test
- [ ] Application test
- [ ] Document test
- [ ] ...

### 5. New dependencies

N/A

### CODE REVIEW

File: python/nano/src/bigdl/nano/pytorch/gpu_cpu/gpu_cpu.py [hunk 1]
```diff
@@ -41,6 +40,14 @@
                       'empty_quantized', 'empty_strided',
                       'frombuffer', 'from_file']
 
+attrs = []
+
+
+def replace_attr(obj, name: str, value):
+    torch_attr = getattr(obj, name)
+    setattr(obj, name, value)
+    attrs.append((obj, name, torch_attr))
```
Comment (TheaperDeng) [inline python/nano/src/bigdl/nano/pytorch/gpu_cpu/gpu_cpu.py:49]: what if `patch_cuda` is called twice? `torch_attr` will be the patched cpu version and the unpatch function will be broken.

File: python/nano/src/bigdl/nano/pytorch/dispatcher.py [hunk 2]
```diff
@@ -91,5 +91,7 @@ def unpatch_torch():
     for mapping_iter in mapping_torch:
         setattr(mapping_iter[0], mapping_iter[1], mapping_iter[3])
 
+    unpatch_cuda()
+
```
Comment (TheaperDeng) [inline python/nano/src/bigdl/nano/pytorch/dispatcher.py:95]: remove the unpatch todo comment?

Comment (MeouSker77) [reply python/nano/src/bigdl/nano/pytorch/dispatcher.py:98]: done

File: python/nano/src/bigdl/nano/pytorch/gpu_cpu/gpu_cpu.py (hunk 1 above)
Comment (MeouSker77) [reply python/nano/src/bigdl/nano/pytorch/gpu_cpu/gpu_cpu.py:50]: fixed it

**Comment by jason-dai** [comment]
Please update the PR description with more details.

**Comment by MeouSker77** [comment]
> Please update the PR description with more details.

done

**Comment by jason-dai** [comment]
> * Use `BIGDL_NANO_PATCH_TORCH` and `BIGDL_NANO_PATCH_CUDA` environment variables to avoid calling patch or unpatch more than once, and telling subprocess whether to call `patch_torch()`.

Is there a way to eliminate the usage of environment variables, which are difficult to manage when you try to deploy the code on a cluster?

**Comment by MeouSker77** [comment]
> Is there a way to eliminate the usage of environment variables, which are difficult to manage when you try to deploy the code in a cluster?

yes, I'll refactor it

File: python/nano/src/bigdl/nano/deps/ray/ray_distributed.py [hunk 3]
```diff
@@ -156,6 +156,8 @@ def _wrapping_function(args_pack: tuple) -> Any:   # type: ignore[override]
         # patch Pytorch and CUDA in subprocess
         if os.environ.get('BIGDL_NANO_PATCH_TORCH') == '1':
             patch_cuda = True if os.environ.get('BIGDL_NANO_PATCH_CUDA') == '1' else False
+            os.environ['BIGDL_NANO_PATCH_TORCH'] = '0'
```
Comment (yangw1234) [inline python/nano/src/bigdl/nano/deps/ray/ray_distributed.py:159]: Are there any places where we document the meaning of `BIGDL_NANO_PATCH_TORCH` and `BIGDL_NANO_PATCH_CUDA`? 
If there are not any, maybe we should add the document.

Comment (MeouSker77) [reply python/nano/src/bigdl/nano/deps/ray/ray_distributed.py:159]: They will only be used internally now, they are just used to indicate if patch has been called and whether subprocess should call patch. I'll eliminate their usage.

**Pull request merged**